actions/actions_whitebox_test.go

```diff
@@ -0,0 +1,176 @@
+package actions
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/fabric8-services/fabric8-wit/gormtestsupport"
+	"github.com/fabric8-services/fabric8-wit/resource"
+	tf "github.com/fabric8-services/fabric8-wit/test/testfixture"
+	"github.com/fabric8-services/fabric8-wit/workitem"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+func TestSuiteAction(t *testing.T) {
+	resource.Require(t, resource.Database)
+	suite.Run(t, &ActionSuite{DBTestSuite: gormtestsupport.NewDBTestSuite()})
+}
+
+type ActionSuite struct {
+	gormtestsupport.DBTestSuite
+}
+
+func createWICopy(wi workitem.WorkItem, state string, boardcolumns []interface{}) workitem.WorkItem {
+	var wiCopy workitem.WorkItem
+	wiCopy.ID = wi.ID
+	wiCopy.SpaceID = wi.SpaceID
+	wiCopy.Type = wi.Type
+	wiCopy.Number = wi.Number
+	wiCopy.Fields = map[string]interface{}{}
+	for k := range wi.Fields {
+		wiCopy.Fields[k] = wi.Fields[k]
+	}
+	wiCopy.Fields["system.state"] = state
+	wiCopy.Fields["system.boardcolumns"] = boardcolumns
+	return wiCopy
+}
+
+func (s *ActionSuite) TestChangeSet() {
+	fxt := tf.NewTestFixture(s.T(), s.DB, tf.CreateWorkItemEnvironment(), tf.WorkItems(2))
+	require.NotNil(s.T(), fxt)
+	require.Len(s.T(), fxt.WorkItems, 2)
+
+	s.T().Run("different ID", func(t *testing.T) {
+		_, err := fxt.WorkItems[0].ChangeSet(*fxt.WorkItems[1])
+		require.NotNil(t, err)
+	})
+
+	s.T().Run("same instance", func(t *testing.T) {
+		changes, err := fxt.WorkItems[0].ChangeSet(*fxt.WorkItems[0])
+		require.Nil(t, err)
+		require.Empty(t, changes)
+	})
+
+	s.T().Run("no changes, same column order", func(t *testing.T) {
+		wiCopy := createWICopy(*fxt.WorkItems[0], "new", []interface{}{"bcid0", "bcid1"})
+		fxt.WorkItems[0].Fields["system.state"] = "new"
```

Comment:
You should really be using the constants here. These names will soon go away and I sure don't want to go over all these manually here.

Reply:
Not in the tests. The tests should fail early. What if there is a typo introduced in the constant? That would then be ignored throughout the tests. I very intentionally use constants in the code but always literals in tests in these cases.

Reply:
Gee, why do we have constants then if we're not using them in tests? Do you expect me to go over all these strings of your manually and change them. This is just not needed.

Reply:
Outdated. Already fixed.